### PR TITLE
FastGRNN CUDA

### PR DESCRIPTION
**Pull request by MJ10**
* Adds a PyTorch CUDA extension `FastGRNNCUDACell` implementing the FastGRNNCell. 
* Setup checks if CUDA is available, if so the extension is built and installed.

### CODE REVIEW

File: pytorch/edgeml_pytorch/cuda/fastgrnn_cuda_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    torch::PackedTensorAccessor<scalar_t,2,torch::RestrictPtrTraits,size_t> nu) {
+  //batch index
+  const int n = blockIdx.y;
+  // column index
+  const int c = blockIdx.x * blockDim.x + threadIdx.x;
+  if (c < pre_comp.size(1)){
+    z_t[n][c] = sigmoid(pre_comp[n][c] + bias_z[n][c]);
+    h_prime_t[n][c] = tanh(pre_comp[n][c] + bias_h_prime[n][c]);
+    
+    new_h[n][c] = (sigmoid(zeta[0][0]) * (1 - z_t[n][c]) + sigmoid(nu[0][0])) * h_prime_t[n][c] + z_t[n][c] * old_h[n][c];
```
Comment (mr-yamraj) [inline pytorch/edgeml_pytorch/cuda/fastgrnn_cuda_kernel.cu]: You can directly pass the sigmoid(zeta[0][0]) and sigmoid(nu[0][0]) to the CUDA kernel instead of computing it again and again on the multiple GPU kernels.

**Comment by mr-yamraj** [comment]
@MJ10 have a look at this [repo](https://dev.azure.com/TheTechTeam/_git/Pytorch%20Hack) 
I too worked on creating the CUDA implementation of FastGRNN 2-3 weeks back to submit it in a PyTorch Hackathon. It has a little more optimized version of what you have implemented and all the gradient checks are also passing.

The time for training only 1 timestep was the same as that of LSTM implementation for 1 timestep.
The time format used below : Forward: minimum/average us | Backward minimum/average us
And the time for training FastGRNN 1 time step using the PyTorch implementation is : Forward: 376.463/491.794 us | Backward 588.179/807.095 us
And the time for training FastGRNN 1 time step using the PyTorch implementation is : Forward: 147.820/193.847 us | Backward 524.998/745.825 us

And when we implement training with unrolling timesteps (multiple timesteps) it was still slower than that of LSTM.
When I tried to track down where the PyTorch is doing the unrolling for multiple timesteps. It turns out that PyTorch uses CUDNN (NVIDIA Developed optimized function for RNN) [link](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp#L781) which handles the unrolling of multiple timesteps.

You can find the documentation of the cudnnRNNForwardTraining function [here](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnRNNForwardTraining) and it looks like they have it optimized for LSTM (not sure about GRU) implementation specifically.

This is all I have got till now, I hope this helps.

**Comment by MJ10** [comment]
mr-yamraj thanks for sharing your implementation. I'll definitely take a look at it.

We also have investigated the performance of cuDNN and are working towards an unrolled version with performance similar to that of cuDNN. This cell is just building towards that.

Thanks for sharing the insights!

**Comment by adityakusupati** [comment]
@MJ10 and @mr-yamraj what version of CUDA are these benchmarked on? Also, what are the datasets you are using for the benchmarking of correctness? Lastly, I see that this supports full rank as of now, how hard is it to support low-rank given it is hard to access and manipulate matrices for a CUDA implemented cells?
The compute graphs look good on the first look and will let you know if I have any other issues. Thanks for the implementation :)
@MJ10, can you share the benchmark numbers (accuracy, timing) vs the LSTMCell and old FastGRNN implementation on pytorch.

**Comment by MJ10** [comment]
@adityakusupati thanks for the feedback. 

I'm using CUDA 10.0. Also the benchmarking we're currently talking about is just in terms of the time for a single forward and backward pass call. Regarding correctness, we're just verifying correctness of the analytical gradient w.r.t the numerical gradient with the [`autograd.gradcheck`](https://pytorch.org/docs/stable/autograd.html#numerical-gradient-checking) method. Both the implementations are based on this [tutorial](https://pytorch.org/tutorials/advanced/cpp_extension.html#writing-a-mixed-c-cuda-extension). [Repo link](https://github.com/pytorch/extension-cpp) for reference. 

I am still working on this implementation, along with an unrolled version. Will update the PR with the benchmarking results (for training on some dataset) as soon as that is done.

**Comment by harsha-simhadri** [comment]
Faster torch/CUDA implementation of FastGRNN cell and RNN. Tested in single layer scenarios, needs more testing on KWS pipeline and multi-layer scenarios.

**Comment by adityakusupati** [comment]
@MJ10 did you get to do any testing? If so can you please let me know regarding the same.

**Comment by MJ10** [comment]
@adityakusupati Thanks for reviewing the code!
We haven't tested on any datasets yet. Will update you with results (speedup and performance) once done. 

Regarding the particular line you mentioned, I agree that we should be passing just the required tensors. I tried to implement this with default arguments, but due to the way PyTorch handles arguments to these custom AutoGrad functions, I wasn't able to figure out a way to implement it that way. 

Also just a clarification about the correctness testing, I compared the outputs and gradients of this CUDA version with the existing Python implementation on 100 random test points, in a single layer as well as stacked configuration.

**Comment by MJ10** [comment]
@adityakusupati Here are some preliminary benchmarking results:
On Googlev2 dataset:

FastGRNNCUDA
300 epochs: Validation Accuracy 89.377
Trained in 5951.07 seconds

FastGRNN
300 epochs: Validation Accuracy 87.632
Trained in 22502.36 seconds

**Comment by adityakusupati** [comment]
I am assuming the 30 class data set. The speedups look good. Thanks

**Comment by MJ10** [comment]
Hi @adityakusupati. the following are numbers for training on Google v2 (35 Classes) along with 50,000 samples from Voxceleb as negative class. 

**300 Epochs (with 250 rolling epochs)**:
FastGRNNCUDA
Validation Accuracy 89.377
Trained in 5951.07 seconds
 
FastGRNN
Validation Accuracy 87.632
Trained in 22502.36 seconds

**50 Epochs without rolling**:
FastGRNNCUDA 
Validation Accuracy 89.458
Trained in 4089.60 seconds
  
FastGRNN 
Validation Accuracy 89.519
Trained in 17271.57 seconds

It is an ~4x speedup.

**Comment by adityakusupati** [comment]
Thanks. Can I have the numbers for the Pytorch internal LSTM/GRU Cells? The speedups look good for FastGRNN. I want to see if we are comparable in training time to LSTM/GRU.
